android-sdk/src/main/java/org/sagebionetworks/bridge/android/manager/UploadManager.java

```diff
@@ -26,76 +30,257 @@
 import okhttp3.RequestBody;
 import retrofit2.Retrofit;
+import rx.Completable;
+import rx.Observable;
 import rx.Single;
+import rx.schedulers.Schedulers;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
 
 /**
  * Created by jyliu on 1/30/2017.
  */
-
 public class UploadManager {
     private static final Logger LOG = LoggerFactory.getLogger(UploadManager.class);
     private static final String CONTENT_TYPE_DATA_ARCHIVE = "application/zip";
 
+    // minimum number of minutes from now an expiration should be
+    private static final int UPLOAD_EXPIRY_WINDOW_MINUTES = 30;
+
     private final ForConsentedUsersApi api;
     private final StudyUploadEncryptor encryptor;
-    private final OkHttpClient s3OkhttpClient;
+    private final UploadDAO uploadDAO;
+    private final OkHttpClient okHttpClient = new OkHttpClient.Builder()
+            .connectTimeout(30, TimeUnit.SECONDS)
+            .readTimeout(30, TimeUnit.SECONDS)
+            .writeTimeout(30, TimeUnit.SECONDS)
+            .retryOnConnectionFailure(false).build();
 
     public UploadManager(AuthenticationManager authenticationManager, StudyUploadEncryptor
-            encryptor, OkHttpClient s3OkhttpClient) {
+            encryptor, UploadDAO uploadDAO) {
         this.api = authenticationManager.getApi();
         this.encryptor = encryptor;
-        this.s3OkhttpClient = s3OkhttpClient;
+        this.uploadDAO = uploadDAO;
     }
 
-    public Single<UploadValidationStatus> getStatus(String uploadId) {
-        return RxUtils.toBodySingle(api.getUploadStatus(uploadId));
+    public static class UploadFile {
+        public String filename;
+        public String contentType;
+        public long fileLength;
+        public String md5Hash;
+        public DateTime createdOn;
     }
 
+    /**
+     * Persists the archive on disk and add it to the queue of pending uploads, runs in an IO thread.
+     *
+     * @param filename filename for the archive
+     * @param archive  archive to be queued
+     * @return information about file produced from the archive
+     */
     @NonNull
-    public Single<UploadValidationStatus> upload(String filename, Archive archive) {
-        Single<UploadFile> uploadFileSingle = Single
-                .fromCallable(() -> persist(filename, archive)).cache();
+    public Single<UploadFile> queueUpload(String filename, Archive archive) {
+        return Single.fromCallable(() -> persist(filename, archive))
+                .subscribeOn(Schedulers.io());
+    }
 
-        Single<UploadSession> uploadSessionSingle = uploadFileSingle
-                .flatMap(this::requestUploadSession);
+    /**
+     * This triggers an attempt to upload all the previously queued upload files
+     * You must call queueUpload before this method will have anything to operate on
+     *
+     * @return Observable with information on if the upload was successful or not
```

Comment:
I did change it to an Observable<Void>, but I changed it back to Completable because Observable<Void> didn't provide an advantage (can't do anything with it). I did some googling on Observable<Void>, Completables, and Observable<Completable> and found a more readable and useful way to run a collection of Completables: Completable.mergeDelayError

(see https://github.com/ReactiveX/RxJava/issues/3807)